tutorials/getting_started/semantic_parsing.md

```diff
@@ -0,0 +1,315 @@
+representations.
+
+We accomplish this with a generic state machine decoder, where the model defines a transition
+system stipulating which actions are valid at any particular state.  If the "actions" are all of
+the tokens in your output vocabulary, and all actions are valid at every state, this would be a
+standard seq2seq model (FOOTNOTE: though a much less efficient one than other ways of doing this).
+By changing what the actions represent, or constraining the valid actions at each timestep of
+decoding, however, we can learn better models that are tailored for our task.  Note also that state
+machines are more general than semantic parsers, and you can use this framework for any structured
+prediction problem that you can formulate as a transition system.
+
+The rest of this tutorial will walk you through the pieces of the semantic parsing framework (TODO:
+make these links):
+
+- Transition functions and training algorithms
+- Keeping track of the decoder's state
+- Defining a transition system (or a logical form language) to specify what actions there are and
+  when they can be taken, and an execution engine for that language
+- Adding context to your model (like a table or a database)
+- Putting the context, the transition system, and the execution engine together into a `World`
+
+## Training a transition function
+
+The fundamental piece of the semantic parsing framework is the `TransitionFunction`.  This is a
+pytorch `Module` that parameterizes state transitions.  That is, given a `State`, the
+`TransitionFunction` returns a scored and ranked list of next `States`:
+
+```python
+class TransitionFunction(torch.nn.Module, Generic[StateType]):
+    def take_step(self,
+                  state: StateType,
+                  max_actions: int = None,
+                  allowed_actions: List[Set] = None) -> List[StateType]:
+        raise NotImplementedError
+```
+
+If you ignore the optional arguments for a minute, this is a simple function that takes a `State`
+of a particular type and returns a list of `States` of the same type.  This function will have some
+parameters, probably based on some kind of LSTM decoder hidden state, that are used to score each
+valid action, construct new `States` for those actions, and return them.  The optional parameters
+are for efficiency, so you don't construct huge numbers of `State` objects if you know you're doing
+a beam search with size `k`, or if you're training and only need to score a few select action
+sequences.
+
+The `TransitionFunction` (along with a `State` object which we'll get to later TODO: add link)
+forms the basis of our decoder state machine, and now we need some way to train it.  There are a
+lot of ways to train state machines, depending on what kind of supervision you have.  You could
+have fully-labeled correct action sequences, a set of possibly correct action sequences, a way to
+check the correctness of finished states, or just a reward function on finished or intermediate
```

Comment:
Examples here illustrating the kinds of supervision you described would be useful.